File: utbot-framework/src/main/java/org/utbot/quickcheck/generator/java/nio/charset/CharsetGenerator.java


package org.utbot.quickcheck.generator.java.nio.charset;

import org.utbot.engine.greyboxfuzzer.util.UtModelGenerator;
import org.utbot.framework.concrete.UtModelConstructor;
import org.utbot.framework.plugin.api.*;
import org.utbot.framework.plugin.api.util.IdUtilKt;
import org.utbot.quickcheck.generator.GenerationStatus;
import org.utbot.quickcheck.generator.Generator;
import org.utbot.quickcheck.random.SourceOfRandomness;

import java.lang.reflect.Method;
import java.nio.charset.Charset;
import java.util.List;

import static java.nio.charset.Charset.availableCharsets;
import static org.utbot.external.api.UtModelFactoryKt.classIdForType;
import static org.utbot.framework.plugin.api.util.IdUtilKt.getStringClassId;

/**
 * Produces values of type {@link Charset}.
 */
public class CharsetGenerator extends Generator<Charset> {
    public CharsetGenerator() {
        super(Charset.class);
    }

    @Override
    public UtModel generate(
            SourceOfRandomness random,
            GenerationStatus status) {

        final UtModelConstructor modelConstructor = UtModelGenerator.getUtModelConstructor();

        final String charsetName = random.choose(availableCharsets().keySet());
        final UtModel charsetNameModel = modelConstructor.construct(charsetName, getStringClassId());

        final Method charsetForName;
        try {
            charsetForName = Charset.class.getMethod("forName", String.class);
        } catch (NoSuchMethodException e) {
            throw new RuntimeException(e);
        }

        final ClassId charsetClassId = classIdForType(Charset.class);
        final ExecutableId charsetForNameId = IdUtilKt.getExecutableId(charsetForName);

        final int modelId = modelConstructor.computeUnusedIdAndUpdate();

        return new UtAssembleModel(
                modelId,
                charsetClassId,
                charsetForNameId.getName() + "#" + modelId,
                new UtExecutableCallModel(null, charsetForNameId, List.of(charsetNameModel)),
                null,
                (a) -> List.of()
        );
    }
}

File: utbot-framework/src/main/kotlin/org/utbot/quickcheck/generator/java/nio/charset/CharsetGenerator.kt
package org.utbot.quickcheck.generator.java.nio.charset

import org.utbot.quickcheck.generator.GeneratorContext
import org.utbot.framework.plugin.api.UtAssembleModel
import org.utbot.framework.plugin.api.UtExecutableCallModel
import org.utbot.framework.plugin.api.UtModel
import org.utbot.framework.plugin.api.util.executableId
import org.utbot.framework.plugin.api.util.id
import org.utbot.framework.plugin.api.util.stringClassId
import org.utbot.quickcheck.generator.GenerationStatus
import org.utbot.quickcheck.generator.Generator
import org.utbot.quickcheck.random.SourceOfRandomness
import java.nio.charset.Charset

/**
 * Produces values of type [Charset].
 */
class CharsetGenerator : Generator(Charset::class.java) {
    override fun generate(
        random: SourceOfRandomness,
        status: GenerationStatus
    ): UtModel {
        val charsetName = random.choose(Charset.availableCharsets().keys)
        val charsetNameModel = generatorContext.utModelConstructor.construct(charsetName, stringClassId)
        val charsetForNameId = Charset::forName.executableId
        val modelId =  generatorContext.utModelConstructor.computeUnusedIdAndUpdate()
        return UtAssembleModel(
            modelId,
            Charset::class.id,
            charsetForNameId.name + "#" + modelId,
            UtExecutableCallModel(null, charsetForNameId, listOf(charsetNameModel)),
        )
    }
}